interpreter/exec/eval.ml

```diff
@@ -441,23 +455,72 @@ let rec step_thread (t : thread) : thread =
         let locals' = List.rev args @ List.map default_value f.it.locals in
         let frame' = {inst = !inst'; locals = List.map ref locals'} in
         let instr' = [Label (n2, [], ([], List.map plain f.it.body)) @@ f.at] in
-        vs', [Frame (n2, frame', ([], instr')) @@ e.at]
+        vs', [Frame (n2, frame', ([], instr')) @@ e.at], No_action
 
       | Func.HostFunc (t, f) ->
-        try List.rev (f (List.rev args)) @ vs', []
+        try List.rev (f (List.rev args)) @ vs', [], No_action
         with Crash (_, msg) -> Crash.error e.at msg
       )
-  in {t with code = vs', es' @ List.tl es}
 
+    | Suspend _, vs ->
+      (* TODO: meaningful timestamp handling *)
+      vs, [e], No_action
+
+  in {t with code = vs', es' @ List.tl es}, action
+
+let rec push_to_inner_stack (c : code) (v : value) : code =
+  let vs, es = c in
+  match es with
+  | {it = Label (n, es0, c'); at} :: es' ->
+    vs, {it = Label (n, es0, push_to_inner_stack c' v); at} :: es'
+  | {it = Frame (n, f, c'); at} :: es' ->
+    vs, {it = Frame (n, f, push_to_inner_stack c' v); at} :: es'
+  | _ ->
+    v :: vs, es
+
+let rec try_unsuspend (c : code) (m : memory_inst) (addr : Memory.address) : code option =
+  let vs, es = c in
+  match es with
+  | {it = Label (n, es0, c'); at} :: es' ->
+    Lib.Option.map (fun c'' -> vs, {it = Label (n, es0, c''); at} :: es') (try_unsuspend c' m addr)
+  | {it = Frame (n, f, c'); at} :: es' ->
+    Lib.Option.map (fun c'' -> vs, {it = Frame (n, f, c''); at} :: es') (try_unsuspend c' m addr)
+  | {it = Suspend (m', addr', timeout); at} :: es' ->
+    if Memory.is_same_memory m m' && addr = addr' then
```

Comment:
```suggestion
    if m == m' && addr = addr' then
```

Reply:
I didn't want to bake in an assumption about structural equality being the "right thing" here (what it it becomes possible to have a two memories with different limits but the same backing store?) - is it ok to keep using the `Memory.same` function (renamed as above)?

Reply:
Hm, not sure that would make sense. Memories have identity in the semantics, as prescribed by allocating an address for them in the store. The use of `==` (which is pointer equality) is an exact reflection of that.

Reply:
Changed.